server.js

```diff
@@ -0,0 +1,269 @@
+    res.redirect("/");
+  }
+);
+
+app.get("/logout", function (req, res) {
+  req.logout();
+  res.redirect("/");
+});
+
+app.get(/.*.(js)/, checkAuthenticated, function (request, response) {
+  file = request.url.substring(1); // here we remove the firs character in the request string which is a "/". this is done because fs gets mad if you dont
+  fileType = "application/javascript"; // this sets the fileType to javascript
+  sendResponse(response, file, fileType);
+});
+app.get(/.*.(css)/, function (request, response) {
+  file = request.url.substring(1); // look at comment above
+  fileType = "text/css"; // sets fileType to css for headers
+  sendResponse(response, file, fileType);
+});
+
+function sendResponse(response, file, fileType) {
+  if (file != "") {
+    fs.readFile(file, (err, data) => {
+      // error handler
+      if (err) {
+        var message = "[ERROR]: " + err;
+        console.log(message); // logs error message to console
+        return404(response); // sends a 404 resource not found to the client
+      } else {
+        // writes a success header
+        response.status(200);
+        response.set({
+          "Content-Type": fileType, // adds content type
+          "Access-Control-Allow-Headers": "*", // for getting around cors rules
+          "Access-Control-Allow-Origin": "*",
+        });
+        response.send(data); // ends the response and sends the data from the file
+        response.end();
+      }
+    });
+  } else {
+    return404();
+  }
+}
+
+function checkAuthenticated(req, res, next) {
+  if (req.isAuthenticated()) {
+    return next();
+  }
+
+  res.redirect("/login");
+}
+
+function checkNotAuthenticated(req, res, next) {
+  if (req.isAuthenticated()) {
+    return res.redirect("/");
+  }
+  next();
+}
+// initialize http server
+console.log("starting server");
+const server = http.createServer(app);
+
+function return404(response) {
+  response.status(404);
+  response.set({
+    "Content-Type": "text/html",
+    "Access-Control-Allow-Headers": "*",
+    "Access-Control-Allow-Origin": "*",
+  });
+  response.send("resource not found");
+  response.end();
+}
+
+// set http server listen ports
+server.listen(PORT, HOST, () => {
+  console.log("[INFO]: http server listening at: " + host + ":" + port);
+});
+
+ipc.serve(SOCKET_PATH, function () {
+  ipc.server.on("connect", () => {
+    console.log("Connected");
+  });
+
+  ipc.server.on("started", (data, socket) => {
+    if (!sockets.has(data.botId)) {
+      console.log("adding socket");
+      sockets.set(data.botId, socket);
+    }
+    console.log(data);
+    broadcast(data);
+  });
+
+  ipc.server.on("data", (data, socket) => {
+    broadcast(data);
+  });
+
+  ipc.server.on("socket.disconnected", function (socket, destroyedSocketID) {
+    ipc.log("client " + destroyedSocketID + " has disconnected!");
+  });
+});
+
+function start(botId) {
+  console.log("[INFO]: New Bot created and started");
+  var username = bots[botId].username;
```

Comment:
can we change this to `bots.botId.username`? Looks a bit nicer and conforms to spec afaik

Reply:
not easy to do. it will happen eventually when i restructure the bot login storage